Bug: Provider crash when modifying policy assignments

It looks like attempting to modify policy assignments at the second level causes a crash, e.g.:
```
policy_assignments_to_modify = {
    corp = {
      policy_assignments = {
        Deny-Public-IP-On-NIC = {
          enforcement_mode = "DoNotEnforce"
        }
      }
    }
  }
```
Results in:
```
module.alz.module.alz_global.data.alz_architecture.this: Still reading... [50s elapsed]

Planning failed. Terraform encountered an error while generating this plan.

╷
│ Error: Plugin did not respond
│
│ The plugin encountered an error, and failed to respond to the plugin6.(*GRPCProvider).ReadDataSource call. The plugin logs may contain more details.
╵

Stack trace from the terraform-provider-alz_v0.15.2.exe plugin:

panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x68 pc=0x1394ac5]

goroutine 171 [running]:
github.com/Azure/alzlib/deployment.(*HierarchyManagementGroup).ModifyPolicyAssignment(0x0, {0xc0005d3a28, 0x15}, 0x0, 0xc0010d63f0, {0x0, 0x0, 0x0}, 0x0, {0x0, ...}, ...)
        github.com/Azure/alzlib@v0.21.7/deployment/managementgroup.go:353 +0x65
github.com/Azure/terraform-provider-alz/internal/provider.modifyPolicyAssignments({_, _}, _, {{0x0, {0x0, 0x0}}, {0x2, {0xc00061d7c0, 0xb}}, {{0x0, ...}, ...}, ...}, ...)
        github.com/Azure/terraform-provider-alz/internal/provider/architecture_data_source.go:178 +0x82b
github.com/Azure/terraform-provider-alz/internal/provider.(*architectureDataSource).Read(0xc0009a2f88, {0x1ba4508, 0xc0007bc750}, {{{{0x1bab390, 0xc000b39bf0}, {0x18aac00, 0xc000ab3440}}, {0x1bae340, 0xc00063f180}}, {{{0x0, ...}, ...}, ...}, ...}, ...)
        github.com/Azure/terraform-provider-alz/internal/provider/architecture_data_source.go:108 +0x7f8
github.com/hashicorp/terraform-plugin-framework/internal/fwserver.(*Server).ReadDataSource(0xc00021a1e0, {0x1ba4508, 0xc0007bc750}, 0xc0006b5280, 0xc000b336d8)
        github.com/hashicorp/terraform-plugin-framework@v1.11.0/internal/fwserver/server_readdatasource.go:103 +0x751
github.com/hashicorp/terraform-plugin-framework/internal/proto6server.(*Server).ReadDataSource(0xc00021a1e0, {0x1ba4508?, 0xc0007bc660?}, 0xc0007bc600)
        github.com/hashicorp/terraform-plugin-framework@v1.11.0/internal/proto6server/server_readdatasource.go:55 +0x389
github.com/hashicorp/terraform-plugin-go/tfprotov6/tf6server.(*server).ReadDataSource(0xc00023a3c0, {0x1ba4508?, 0xc0008d7860?}, 0xc00011cb40)
        github.com/hashicorp/terraform-plugin-go@v0.23.0/tfprotov6/tf6server/server.go:688 +0x26d
github.com/hashicorp/terraform-plugin-go/tfprotov6/internal/tfplugin6._Provider_ReadDataSource_Handler({0x19e1300, 0xc00023a3c0}, {0x1ba4508, 0xc0008d7860}, 0xc00042a700, 0x0)      
        github.com/hashicorp/terraform-plugin-go@v0.23.0/tfprotov6/internal/tfplugin6/tfplugin6_grpc.pb.go:572 +0x1a6
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000191000, {0x1ba4508, 0xc0008d7740}, {0x1bacc00, 0xc0003f6000}, 0xc00075b440, 0xc000319110, 0x22755a0, 0x0)
        google.golang.org/grpc@v1.66.0/server.go:1393 +0xe11
google.golang.org/grpc.(*Server).handleStream(0xc000191000, {0x1bacc00, 0xc0003f6000}, 0xc00075b440)
        google.golang.org/grpc@v1.66.0/server.go:1804 +0xe8b
google.golang.org/grpc.(*Server).serveStreams.func2.1()
        google.golang.org/grpc@v1.66.0/server.go:1029 +0x7f
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 36
        google.golang.org/grpc@v1.66.0/server.go:1040 +0x125

Error: The terraform-provider-alz_v0.15.2.exe plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.

```

However when modifying a policy one level higher, there's no issue.
```
policy_assignments_to_modify = {
    landing_zones = {
      policy_assignments = {
        Enable-DDoS-VNET = {
          enforcement_mode = "DoNotEnforce"
        }
      }
    }
  }
```

--- a/internal/provider/architecture_data_source.go
+++ b/internal/provider/architecture_data_source.go
@@ -150,6 +150,14 @@ func (d *architectureDataSource) Read(ctx context.Context, req datasource.ReadRe
 
 func modifyPolicyAssignments(ctx context.Context, depl *deployment.Hierarchy, data gen.ArchitectureModel, resp *datasource.ReadResponse) {
 	for mgName, pa2modValue := range data.PolicyAssignmentsToModify.Elements() {
+		mg := depl.ManagementGroup(mgName)
+		if mg == nil {
+			resp.Diagnostics.AddWarning(
+				"architectureDataSource.Read() Warning modifying policy assignments",
+				fmt.Sprintf("Management group `%s` not found in hierarchy", mgName),
+			)
+			return
+		}
 		pa2mod, ok := pa2modValue.(gen.PolicyAssignmentsToModifyValue)
 		if !ok {
 			resp.Diagnostics.AddError(
@@ -175,7 +183,7 @@ func modifyPolicyAssignments(ctx context.Context, depl *deployment.Hierarchy, da
 				)
 				return
 			}
-			if err := depl.ManagementGroup(mgName).ModifyPolicyAssignment(paName, params, enf, noncompl, ident, resourceSel, overrides); err != nil {
+			if err := mg.ModifyPolicyAssignment(paName, params, enf, noncompl, ident, resourceSel, overrides); err != nil {
 				resp.Diagnostics.AddError(
 					"architectureDataSource.Read() Error modifying policy assignment values in alzlib",
 					fmt.Sprintf("Error modifying policy assignment values for `%s` at mg `%s`: %s", paName, mgName, err.Error()),

--- a/internal/provider/architecture_data_source_test.go
+++ b/internal/provider/architecture_data_source_test.go
@@ -29,7 +29,7 @@ func TestAccAlzArchitectureDataSourceRemoteLib(t *testing.T) {
 		ExternalProviders: map[string]resource.ExternalProvider{
 			"azapi": {
 				Source:            "azure/azapi",
-				VersionConstraint: "~> 1.14",
+				VersionConstraint: "~> 2.0",
 			},
 		},
 		Steps: []resource.TestStep{
@@ -52,7 +52,7 @@ func TestAccAlzArchitectureDataSourceWithDefaultAndModify(t *testing.T) {
 		ExternalProviders: map[string]resource.ExternalProvider{
 			"azapi": {
 				Source:            "azure/azapi",
-				VersionConstraint: "~> 1.14",
+				VersionConstraint: "~> 2.0",
 			},
 		},
 		Steps: []resource.TestStep{
@@ -87,7 +87,7 @@ func TestAccAlzArchitectureDataSourceExistingMg(t *testing.T) {
 		ExternalProviders: map[string]resource.ExternalProvider{
 			"azapi": {
 				Source:            "azure/azapi",
-				VersionConstraint: "~> 1.14",
+				VersionConstraint: "~> 2.0",
 			},
 		},
 		Steps: []resource.TestStep{
@@ -99,6 +99,24 @@ func TestAccAlzArchitectureDataSourceExistingMg(t *testing.T) {
 				Check: resource.ComposeAggregateTestCheckFunc(
 					resource.TestCheckResourceAttr("data.alz_architecture.test", "id", "existingmg"),
 				),
+			},
+		},
+	})
+}
+
+func TestAccAlzArchitectureDataSourceModifyPolicyAssignmentNonExistent(t *testing.T) {
+	resource.Test(t, resource.TestCase{
+		PreCheck:                 func() { testAccPreCheck(t) },
+		ProtoV6ProviderFactories: testAccProtoV6ProviderFactoriesUnique(),
+		ExternalProviders: map[string]resource.ExternalProvider{
+			"azapi": {
+				Source:            "azure/azapi",
+				VersionConstraint: "~> 2.0",
+			},
+		},
+		Steps: []resource.TestStep{
+			{
+				Config: testAccArchitectureDataSourceConfigModifyPolicyAssignmentNonExistent(),
 			},
 		},
 	})
@@ -273,6 +291,35 @@ data "alz_architecture" "test" {
 
 output "management_group_exists" {
 	value = data.alz_architecture.test.management_groups[0].exists
+}
+`
+}
+
+func testAccArchitectureDataSourceConfigModifyPolicyAssignmentNonExistent() string {
+	return `
+provider "alz" {
+  library_references = [
+    {
+	    custom_url = "${path.root}/testdata/testacc_lib"
+	  }
+	]
+}
+
+data "azapi_client_config" "current" {}
+
+data "alz_architecture" "test" {
+  name                     = "test"
+  root_management_group_id = data.azapi_client_config.current.tenant_id
+  location                 = "swedencentral"
+  policy_assignments_to_modify = {
+    not_exist = {
+      policy_assignments = {
+        Deploy-MDEndpoints = {
+          enforcement_mode = "DoNotEnforce"
+        }
+      }
+    }
+  }
 }
 `
 }

